Mathlib/Topology/CWComplex/Basic.lean

```diff
@@ -0,0 +1,682 @@
+/-
+Copyright (c) 2024 Floris van Doorn and Hannah Scholz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn, Hannah Scholz
+-/
+
+import Mathlib.Analysis.InnerProductSpace.Basic
+import Mathlib.Data.Real.StarOrdered
+import Mathlib.Order.CompletePartialOrder
+
+/-!
+# CW-complexes
+
+This file defines (relative) CW-complexes and proofs basic properties about them.
+
+A CW-complex is a topological space that is made by gluing closed disks of different dimensions
+together.
+
+## Main definitions
+* `RelCWComplex C D` : The class of CW-structures on a subspace `C` realtive to `D` of a topological
+  space `X`.
+* `ClasCWComplex C` : An abbreviation for `RelCWComplex C ∅`. The class of CW-structures on a
+  subspace `C` of the topological space `X`.
+* `openCell n i` : An open cell of dimension `n`.
+* `closedCell n i` : A closed cell of dimension `n`.
+* `cellFrontier n i` : The edge of a cell of dimension `n`.
+* `level C n` : The `n`-th level of the (relative) CW-complex `C`.
+
+## Main statements
+* `iUnion_openCell_eq_level` : The levels can also be seen as a union of open cells.
+* `cellFrontier_subset_finite_openCell` : The edge of a cell is contained in a finite union of
+  open cells of a lower dimension.
+
+## Implementation notes
+* This definition uses is the historical definition rather than the modern one to avoid working
+  with a lot of different topological spaces.
+* We define what it means for a subspace to be a CW-complex instead of a topological space itself.
+  This has the advantage that it makes a lot of constructions easier as you do not need to deal with
+  different topologies. However, this approach requires the subspace to be closed. Should that
+  not be the case you need to consider that subspace as a subspace of itself.
+* The definition `RelCWcomplex` does not require `X` to be a Hausdorff space.
+  A lot of the lemmas will however require this property.
+* For statements the auxiliary construction `levelaux` is preferred over `level` as it makes
+  the base case of inductions easier. The statement about `level` should then be derived from the
+  one about `levelaux`.
+
+## References
+* [A. Hatcher, *Algebraic Topology*][hatcher02]
+-/
+
+noncomputable section
+
+open Metric Set
+
+/-- Characterizing when a subspace `C` of a topological space `X` is a CW-complex relative to
+  another subspace `D`. Note that this requires `C` and `D` to be closed subspaces.
+  If `C` is not closed choose `X` to be `C`. A lot of lemmas will require `[T2Space X]`.-/
```

Comment:
I would just remove "A lot of lemmas will require `[T2Space X]`." I'm not sure what information this adds. What do you think?